juju/juju
# Operations persistence

## Description of change

Add "operations" to the juju persistent data model.
Operations group together actions run against multiple units - each action invocation on a unit is a task. There 2 commits:
1. the core operations persistence code
2. upgrade steps

The main work items are:
- add operations to mongo state
- update parent operation status / completion time etc whenever a child task/action is updated
- adjust pruning so that operations are pruned, not actions

The actions code to Begin()/Finish() an action wasn't using the buildTxn pattern so that all needed updating to properly implement checks for race conditions when updating an operation.

Pruning was adjusted so that for prune by size, the actions collection is what is used to measure the desired size (actions can be large), but the pruning is done on operations. A ratio of operation count to action count is used to estimate how many operations need to be pruned to reduce the actions collection size. For large counts, this will suffice.

## QA steps

Operations are not exposed yet. Test by juju running an action across multiple units and using the mongo client to look at the operations collection to see the mongo doc for the operation.

Next PR will start to wire the operations up to the CLI.
